### DISCUSSION TITLE
Persistence of custom plugin state?

### DISCUSSION THREAD

**Discussion by gryphonmyers** [beetbox/beets#4817] [Q&A] [answered]
I can't find any documentation or example plugins that do this, so if the answer is obvious, I apologize in advance. I'm thinking about writing a plugin that needs to persist some custom state (musicbrainz ids that are determined based on a combination of API calls and beets library queries). It does not make sense to extend the beets db models for this purpose - it is plugin-specific state that does not pertain to existing library items or albums. What is considered best practice for storage and retrieval of this kind of data?

**Comment by sampsyo** [comment, marked as answer]
We actually do not have a general-purpose way to do this! It's common to attach attributes to items or albums for plugin-specific reasons, but we don't have a way that's unattached to the models.

You might consider taking inspiration from the way the importer saves its progress state, which is just storing a pickled data structure in the configuration directory:
https://github.com/beetbox/beets/blob/0c3f428a601cb40c5fd463791df6229d51b0635e/beets/importer.py#L69-L91
